xla/service/gpu/fusions/custom.cc

```diff
@@ -812,10 +813,136 @@ absl::StatusOr<FusionEmissionResult> CustomFusion::Emit(
+  std::vector<std::optional<uint64_t>> offset_byte_sizes(2, std::nullopt);
+
+  std::vector<HloInstruction*> slice_instrs(2, nullptr);
+  std::vector<std::optional<BufferAllocation::Slice>> arguments;
+
+  // Collect slice information for inputs.
+  unsigned arg_idx = 0;
+  TF_ASSIGN_OR_RETURN(arguments.emplace_back(),
+                      GetOperandSlice(buffer_assignment, adaptor, fusion_instr,
+                                      *instr->operand(arg_idx), slice_instrs,
+                                      /*shape_idx=*/{}, arg_idx));
+  TF_RETURN_IF_ERROR(CollectSliceInfo(
+      buffer_assignment, fusion_instr,
+      absl::Span<HloInstruction*>(slice_instrs), offset_buffer_indices,
+      orig_shapes, sliced_shapes, offset_byte_sizes, arg_idx++));
+
+  // Collect slice information for outputs.
+  TF_ASSIGN_OR_RETURN(
+      arguments.emplace_back(),
+      GetResultSlice(buffer_assignment, adaptor, fusion_instr, *instr,
+                     slice_instrs, /*shape_idx=*/{}, arg_idx));
+  TF_RETURN_IF_ERROR(CollectSliceInfo(
+      buffer_assignment, fusion_instr,
+      absl::Span<HloInstruction*>(slice_instrs), offset_buffer_indices,
+      orig_shapes, sliced_shapes, offset_byte_sizes, arg_idx));
+
+  if (absl::c_all_of(slice_instrs, [&](HloInstruction* slice_instr) {
+        return slice_instr &&
+               slice_instr->opcode() != HloOpcode::kDynamicUpdateSlice;
+      })) {
+    return absl::InternalError(
+        "DynamicSliceFusion with reduce-scatter expects a dynamic-update-slice "
+        "operation.");
+  }
+
+  // Provide fake allocations for inputs and outputs.
+  std::vector<std::unique_ptr<BufferAllocation>> fake_allocations(2);
+  unsigned fake_arg_idx = 0;
+  int64_t operand_byte_size =
+      ShapeUtil::ByteSizeOf(instr->operand(fake_arg_idx)->shape());
+  fake_allocations[fake_arg_idx] = std::make_unique<BufferAllocation>(
+      /*index=*/fake_arg_idx, operand_byte_size, /*color=*/0);
+  BufferAllocation::Slice slice_operand(fake_allocations[fake_arg_idx].get(), 0,
+                                        operand_byte_size);
+  fake_arg_idx++;
+  TF_RET_CHECK(instr->shape().IsArray() &&
+               "The output is not expected to be a tuple.");
+  int64_t out_fake_byte_size =
+      ShapeUtil::ByteSizeOf(instr->shape());  // TODO: we don't need this
+  fake_allocations[fake_arg_idx] = std::make_unique<BufferAllocation>(
+      /*index=*/fake_arg_idx, out_fake_byte_size, /*color=*/0);
+  BufferAllocation::Slice slice_out_fake(fake_allocations[fake_arg_idx].get(),
+                                         0, out_fake_byte_size);
+
+  // Generate the hero thunk and wrap it in a dynamic-slice thunk.
+  ThunkSequence seq;
+  auto thunk_info = Thunk::ThunkInfo::WithProfileAnnotation(instr);
+  std::vector<NcclCollectiveThunk::Buffer> buffers;
+  const Shape& src_shape = instr->operand(0)->shape();
+  const Shape& dst_shape = instr->shape();
+  buffers.push_back(NcclCollectiveThunk::Buffer{
```

Comment:
Please don't use C++20: Designated initializers are incompatible with C++ standards before C++20

Reply:
Done. Thanks for pointing this out.